wicket/src/update.rs

```diff
@@ -0,0 +1,241 @@
+    where
+        S: serde::Serializer,
+    {
+        hex::serde::serialize(self.0, serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Sha3_256Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(Sha3_256Digest(hex::serde::deserialize(deserializer)?))
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum ArtifactType {
+    // Sled Artifacts
+    SledSp,
+    SledRoT,
+    HostPhase1,
+    HostPhase2,
+
+    // PSC Artifacts
+    PscSp,
+    PscRot,
+
+    // Switch Artifacts
+    SwitchSp,
+    SwitchRot,
+}
+
+/// A description of a software artifact that can be installed on the rack
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Artifact {
+    filename: PathBuf,
+    artifact_type: ArtifactType,
+    version: Version,
+    digest: Sha3_256Digest,
+    length: u64,
+}
+
+/// Attempt to convert an ArtifactSpec to an Artifact
+/// by reading from the filesysetm and hashing.
+impl TryFrom<ArtifactSpec> for Artifact {
+    type Error = std::io::Error;
+    fn try_from(spec: ArtifactSpec) -> Result<Self, Self::Error> {
+        let mut hasher = Sha3_256::new();
+        let mut file = File::open(&spec.filename)?;
+        let length = std::io::copy(&mut file, &mut hasher)?;
+        let mut digest = [0u8; 32];
+        digest.copy_from_slice(&hasher.finalize());
+        let digest = Sha3_256Digest(digest);
+        Ok(Artifact {
+            filename: spec.filename,
+            artifact_type: spec.artifact_type,
+            version: spec.version,
+            digest,
+            length,
+        })
+    }
+}
+
+/// User Input that describes artifacts as part of the [`RackUpdateSpec`]
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ArtifactSpec {
+    filename: PathBuf,
+    artifact_type: ArtifactType,
+    version: Version,
+}
+
+/// The set of all artifacts in an update
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Manifest {
+    version: Version,
+    artifacts: Vec<Artifact>,
+}
+
+impl Manifest {
+    pub fn dump<P: AsRef<Path>>(
+        &self,
+        path: P,
+    ) -> Result<PathBuf, std::io::Error> {
+        let json = serde_json::to_string(&self)?;
+        let path = path.as_ref().join("manifest.json");
+        let mut file = File::create(&path)?;
+        file.write_all(json.as_bytes())?;
+        Ok(path)
+    }
+
+    /// Unpack a Tar archive into `unpack_dir`, read the manifest, and return
+    /// the manifest.
+    pub fn load(
+        tarfile: impl AsRef<Path>,
+        unpack_dir: impl AsRef<Path>,
+    ) -> std::io::Result<Manifest> {
+        let tarfile = File::open(tarfile)?;
+        let mut archive = tar::Archive::new(tarfile);
+        archive.unpack(&unpack_dir)?;
+        let json = std::fs::read_to_string(
+            &unpack_dir.as_ref().join("manifest.json"),
+        )?;
+        let manifest = serde_json::from_str(&json)?;
```

Comment:
Similar to above, I think we could skip reading the file contents into a temporary string:

```rust
let file = File::open(unpack_dir.as_ref().join("manifest.json"))?;
let manifest = serde_json::from_reader(file)?;
```